zipline/gens/tradesimulation.py

```diff
@@ -100,6 +101,16 @@ def transform(self, stream_in):
                 self.simulation_dt = date
                 self.on_dt_changed(date)
 
+                # removing expired futures
+                for sid in self.current_data.keys():
+                    try:
+                        if self.env.asset_finder.retrieve_asset(sid).end_date \
+                                < self.simulation_dt:
+                            self.current_data.__delitem__(sid)
+                    except (AttributeError, TypeError, ValueError,
```

Comment:
> they are sids.

:(

>  I agree but I'm not sure how else the expired data can be removed.

One possibility would be to pre-compute the sids that will expire on each backtest day, and then only loop over those on each day.  A naive way to do this would be to build such a lookup table using all assets in our asset finder, and a more refined approach would be to only build the lookup table for assets that will be emitted by our DataSource, with the caveat that the set of emittable sids may not be knowable in advance for dynamic sources.